Source/WebCore/Modules/WebGPU/GPUQueue.cpp

```diff
@@ -328,16 +329,182 @@ static void imageBytesForSource(const auto& source, const auto& destination, Ima
+    }, [&](const RefPtr<ImageData> imageData) -> ResultType {
+        auto width = imageData->width();
+        auto height = imageData->height();
+        if (width < 0 || height < 0 || horizontalDimension > static_cast<uint32_t>(width) || verticalDimension > static_cast<uint32_t>(height)) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        auto size = width * height;
+        if (!size) {
+            errorCode = ExceptionCode::InvalidStateError;
+            return false;
+        }
+        return true;
+    }, [&](const RefPtr<HTMLImageElement> imageElement) -> ResultType {
+        if (!imageElement->cachedImage()) {
+            errorCode = ExceptionCode::InvalidStateError;
+            return false;
+        }
+        if (horizontalDimension > imageElement->width() || verticalDimension > imageElement->height()) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        return true;
+    }, [&](const RefPtr<HTMLVideoElement>) -> ResultType {
+        return true;
+    }, [&](const RefPtr<WebCodecsVideoFrame>) -> ResultType {
+        return true;
+#endif
+    }, [&](const RefPtr<HTMLCanvasElement>& canvas) -> ResultType {
+        if (!canvas->renderingContext()) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        if (canvas->renderingContext()->isPlaceholder()) {
+            errorCode = ExceptionCode::InvalidStateError;
+            return false;
+        }
+        if (horizontalDimension > canvas->width() || verticalDimension > canvas->height()) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        return true;
+    }
+#if ENABLE(OFFSCREEN_CANVAS)
+    , [&](const RefPtr<OffscreenCanvas>& offscreenCanvas) -> ResultType {
+        if (offscreenCanvas->isDetached()) {
+            errorCode = ExceptionCode::InvalidStateError;
+            return false;
+        }
+        if (!offscreenCanvas->renderingContext()) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        if (offscreenCanvas->renderingContext()->isPlaceholder()) {
+            errorCode = ExceptionCode::InvalidStateError;
+            return false;
+        }
+        if (horizontalDimension > offscreenCanvas->width() || verticalDimension > offscreenCanvas->height()) {
+            errorCode = ExceptionCode::OperationError;
+            return false;
+        }
+        return true;
+    }
+#endif
+    );
+}
+
 // FIXME: https://bugs.webkit.org/show_bug.cgi?id=263692 - this code should be removed, it is to unblock
 // compiler <-> pipeline dependencies
-static void* copyToDestinationFormat(const uint8_t* rgbaBytes, GPUTextureFormat format, size_t& sizeInBytes)
+#if PLATFORM(COCOA)
+static uint32_t convertRGBA8888ToRGB10A2(uint8_t r, uint8_t g, uint8_t b, uint8_t a)
 {
+    uint32_t r0 = static_cast<uint32_t>(r) * 4 + 1;
```

Comment:
Is this right? Eg, 255 will be converted to 1021, which is just a little bit smaller than 1023, such that 255/255 == 1.0f != 1021/1023.

I would expect that a correct conversion would keep the same float representation (as near as possible).

Reply:
Oh indeed. I’ll change this to:

`(c` * 1023) / 255`

instead, thanks for catching

Reply:
thanks for catching @djg I fixed the alpha calculation which was really wrong as well

Reply:
If you want to avoid the `mul` and `div` you might be able to get away with copying the 2 msbs from the source into the 2 lsbs of the dest. ie:

```
r0 = ((uint32_t) r << 2) | ((uint32_t) r >> 6)
```

Reply:
My ex-boss, Charles, it probably the authority on the issues involved. http://cbloomrants.blogspot.com/2021/04/requantization-to-different-unorm-bit.html

Reply:
I'll take that change :) The values are effectively the same and in some cases the bit shifting might be more accurate as it emulates rounding - https://onlinegdb.com/dKdKd1gzo